### DISCUSSION TITLE
micropython/lora, how to do TX/RX switching??

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#13262] [Libraries & Drivers]
I have separate TX and RX working on two devices.  Now, I am testing running TX and RX on both devices and running into a problem.

It appears that once you call `receiver = Receiver(modem)` then you just can't call `sender = Sender(modem, DEVICE_ID)` or you are told 
`RuntimeError: Receiving`  

I looked about for a `modem.deinit()` and don't see anything.

How would one properly switch between TX and RX?

Thank you.

**Comment by sophiedegran** [comment]
Daves did you try with fix transmission monitoring and then filter on received address ?

**Reply by davefes** [reply]
I have got the separate "reliable_delivery" examples in `micropython/lora` running.  In a previous distro:
https://github.com/Wei1234c/SX127x_driver_for_MicroPython_on_ESP8266/tree/master/examples/DuplexCallback
it was straight-forward getting two units talking to each other.

I suspect rather than using `sender.py` and `receiver.py` in the examples at `micropython/lora` that I may have construct the application myself.

**Reply by sophiedegran** [reply]
Dave, I use ebyte lora library for sx127x that's very good improved library, have a look to mischianti.org
You will see how to manage tx/rx on the same lora card ..

**Reply by davefes** [reply]
@sophiedegran There is a lot of good information at mischianti.org

I have invested quite a bit of time in learning `micropython/lora` so would really like to use it for this application.  Thank you for the hint.

**Reply by davefes** [reply]
:
@sophiedegran Don't see anything TX/RX-related in the ebyte lora examples. 

What is interesting in this distro is that he is using the UART to communicate with the board.  I had go at trying to get the E77-400M22S to work, but had problems uploading files to the devices via UART ... maybe there is something helpful in mischianti's distro.

Edit:  I don't see the keywords `ack` or `retries`.  `micropython/lora` handles `acks` and `retries`, which is 2 steps further than:
https://github.com/Wei1234c/SX127x_driver_for_MicroPython_on_ESP8266/tree/master/

**Comment by davefes** [comment]
I would like to send a "reliable_delivery" response when I get a message in a receive callback().

It seems that when you register a `set_irq_callback()` in the main program loop and one gets a received message that program stays in the callback.  This means other processing can not take place.  Is there some way to get back to the main processing loop.

Doesn't return, but continues to receive messages:
```py
def my_rx_irq(receiver):
    my_data = []


  # With wait=True, this function blocks until something is received and always
  # returns non-None
    sender_id, data = receiver.recv(wait=True)

    data =  data.decode('utf-8')
    my_data = data.split(',')
    print (f'Received msg: {my_data[0]}\n')
```

**Comment by davefes** [comment]
Gave-up and moved to using the "low-level API"
